src/twisted/application/runner/_pidfile.py

```diff
@@ -0,0 +1,202 @@
+# See LICENSE for details.
+
+"""
+PID file.
+"""
+
+import errno
+from os import getpid, kill
+
+from twisted.python.filepath import IFilePath
+from twisted.logger import Logger
+
+
+
+class PIDFile(object):
+    """
+    PID file.
+
+    Manages a file that remembers a process ID.
+    """
+
+    _log = Logger()
+
+
+    @staticmethod
+    def format(pid=None):
+        """
+        Format a PID file's content.
+
+        @param pid: A process ID.
+        @type pid: int
+
+        @return: Formatted PID file contents.
+        @rtype: L{bytes}
+        """
+        if pid is not None:
+            return u"{}\n".format(pid).encode("utf-8")
+        else:
+            return b""
+
+
+    def __init__(self, filePath):
+        """
+        @param filePath: The path to the PID file on disk.
+        @type filePath: L{IFilePath}
+        """
+        self.filePath = IFilePath(filePath)
+
+
+    def read(self):
+        """
+        Read the process ID stored in this PID file.
+
+        @return: The contained process ID, or C{None} if the PID file does not
+            exist.
+        @rtype: L{int}
+
+        @raise EnvironmentError: If this PID file cannot be read.
+        @raise ValueError: If this PID file's content is invalid.
+        """
+        pidString = b""
+        try:
+            for pidString in self.filePath.open():
+                break
+        except OSError as e:
+            if e.errno == errno.ENOENT:  # No such file
+                return None
+            raise
+
+        return int(pidString)
+
+
+    def write(self, pid=None):
+        """
+        Store a PID in this PID file.
+
+        @param pid: A PID to store.  If C{None}, store the currently running
+            process ID.
+        @type pid: L{int}
+
+        @raise EnvironmentError: If this PID file cannot be written.
+        """
+        if pid is None:
+            pid = getpid()
+        else:
+            pid = int(pid)
+
+        self.filePath.setContent(self.format(pid=pid))
+
+
+    def remove(self):
+        """
+        Remove this PID file.
+
+        @raise EnvironmentError: If this PID file cannot be removed.
+        """
+        self.filePath.remove()
+
+
+    def isRunning(self):
+        """
+        Determine whether there is a running process corresponding to the PID
+        in this PID file.
+
+        @return: True if this PID file contains a PID and a process with that
+        PID is currently running; false otherwise.
+        @rtype: L{bool}
+
+        @raise EnvironmentError: If this PID file cannot be read.
+        @raise ValueError: If this PID file's content is invalid.
+        """
+        pid = self.read()
+        if pid is None:
+            return False
+
+        try:
+            kill(pid, 0)
+        except OSError as e:
+            if e.errno == errno.ESRCH:  # No such process
+                self._log.info("Removing stale PID file: {log_source}")
+                self.remove()
```

Comment:
Potentially weird side-effect: a boolean function modifying global (filesystem) state. I realize an invalid PID is equivalent to an empty file in some respect, but this feels like it can cause hard-to-debug problems.